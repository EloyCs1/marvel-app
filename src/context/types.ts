import { ReactNode } from "react";

export interface FavoriteContextValue {
  favoriteFilter: boolean;
  toggleFavoriteFilter: () => void;
  favoriteIds: number[];
  toggleFavoriteId: (id: number) => void;
}

export interface FavoriteProviderProps {
  children: ReactNode;
  values: FavoriteContextValue;
}
