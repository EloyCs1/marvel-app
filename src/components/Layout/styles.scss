.layout {
  &__content {
    margin-top: 84px;
    overflow-y: auto;
    height: calc(100vh - 84px);
  }
}

.navbar {
  position: fixed;
  width: -webkit-fill-available;
  top: 0;
  z-index: 1;
  height: 84px;
  background-color: #000;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 0 52px 0 48px;

  a {
    height: 52px;
  }

  &__favorites {
    display: flex;
    gap: 8px;
    padding: 8px;
    align-items: center;

    p {
      color: var(--color-white);
    }
  }

  @media (max-width: 575px) {
    padding: 0 16px 0 16px;
  }
}
