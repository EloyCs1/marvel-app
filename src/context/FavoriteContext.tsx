import { createContext, useCallback, useContext, useState } from "react";
import { FavoriteContextValue, FavoriteProviderProps } from "./types";

export const defaultFavoriteContextValue: FavoriteContextValue = {
  favoriteFilter: false,
  toggleFavoriteFilter: () => console.warn("toggleFavoriteFilter init"),
  favoriteIds: [],
  toggleFavoriteId: () => console.warn("toggleFavoriteId init"),
};

export const FavoriteContext = createContext(defaultFavoriteContextValue);

export const FavoriteProvider = ({ children, values }: FavoriteProviderProps) => {
  const [favoriteFilter, setFavoriteFilter] = useState(values.favoriteFilter);
  const [favoriteIds, setFavoriteIds] = useState(values.favoriteIds);

  const toggleFavoriteFilter = useCallback((value: boolean) => setFavoriteFilter(value), []);
  const toggleFavoriteId = useCallback(
    (id: number) => {
      const index = favoriteIds.indexOf(id);
      if (index > -1) {
        const ids = [...favoriteIds];
        ids.splice(index, 1);
        setFavoriteIds(ids);
        return;
      }
      setFavoriteIds([...favoriteIds, id]);
    },
    [favoriteIds],
  );

  return (
    <FavoriteContext.Provider value={{ favoriteFilter, toggleFavoriteFilter, favoriteIds, toggleFavoriteId }}>
      {children}
    </FavoriteContext.Provider>
  );
};

export const useFavorites = () => useContext(FavoriteContext);
