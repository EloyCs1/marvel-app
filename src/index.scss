:root {
  --color-red-dark: #a51419;
  --color-red-light: #ef4a5044;
  --color-red: #ec1d24;
  --color-black: #000000;
  --color-white: #ffffff;
  --color-border: #333333;
  --color-disabled: #bfbfbf;
}

body {
  margin: 0;
  padding: 0;
  font-family:
    Roboto Condensed,
    -apple-system,
    BlinkMacSystemFont,
    "Segoe UI",
    "Roboto",
    "Oxygen",
    "Ubuntu",
    "Cantarell",
    "Fira Sans",
    "Droid Sans",
    "Helvetica Neue",
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-size: 16px;
  font-weight: 400;
  line-height: 18.75px;
  text-align: left;
  color: var(--color-black);
}

p {
  margin: 0;
  padding: 0;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New", monospace;
}
