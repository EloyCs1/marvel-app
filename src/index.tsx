import React from "react";
import ReactDOM from "react-dom/client";
import "./index.scss";
import AppRouter from "./pages/AppRouter";
import reportWebVitals from "./reportWebVitals";
import { IntlProvider } from "react-intl";
import messages, { DEFAULT_LOCALE, getLocales } from "./lang";

const root = ReactDOM.createRoot(document.getElementById("root") as HTMLElement);

const navigatorLanguage = navigator.language.split(/[-_]/)[0];
const locale = getLocales().includes(navigatorLanguage) ? navigatorLanguage : DEFAULT_LOCALE;

root.render(
  <React.StrictMode>
    <IntlProvider locale={locale} messages={messages[locale as keyof typeof messages]} defaultLocale={DEFAULT_LOCALE}>
      <AppRouter />
    </IntlProvider>
  </React.StrictMode>,
);

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals();
