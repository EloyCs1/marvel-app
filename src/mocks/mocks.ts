import { AppCharacter } from "../service/types";

export const CHARACTER: AppCharacter.Character = {
  description: "Rick Jones has been Hulk's best bud since day one ",
  id: 1017100,
  image: "http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16.jpg",
  name: "A-Bomb (HAS)",
};
