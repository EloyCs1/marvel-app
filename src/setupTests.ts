// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";

export const mockHistoryPush = jest.fn();
export const mockHistoryGoBack = jest.fn();
export const mockHistoryReplace = jest.fn();
export const mockLocation = jest.fn();
export const mockNavigate = jest.fn();
export const mockUseParams = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
    replace: mockHistoryReplace,
    goBack: mockHistoryGoBack,
  }),
  useLocation: () => mockLocation,
  useParams: mockUseParams,
  useNavigate: () => mockNavigate,
}));
