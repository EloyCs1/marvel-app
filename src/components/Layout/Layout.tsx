import React from "react";

import { Link, useNavigate } from "react-router-dom";
import Icon from "src/components/common/Icon/Icon";
import { IconName } from "src/components/common/Icon/types";
import { useFavorites } from "src/context/FavoriteContext";
import { LayoutProps } from "./types";
import "./styles.scss";

const Layout: React.FC<LayoutProps> = ({ children }) => {
  const testId = "layout";
  const navigate = useNavigate();
  const { favoriteIds, toggleFavoriteFilter } = useFavorites();

  const onClickLogo = () => {
    toggleFavoriteFilter(false);
  };

  const onClickIcon = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.preventDefault();
    e.stopPropagation();
    toggleFavoriteFilter(true);
    navigate("/");
  };

  return (
    <>
      <div className="navbar" data-testid={`${testId}-navbar`}>
        <Link data-testid={`${testId}-link`} to={"/"} onClick={onClickLogo}>
          <Icon testId={`${testId}-logo`} width={130} height={52} name={IconName.MARVEL_LOGO} />
        </Link>
        <div className="navbar__favorites">
          <Icon testId={`${testId}-icon`} name={IconName.FAVORITES_FILLED} onClick={onClickIcon} />
          <p data-testid={`${testId}-count`}>{favoriteIds.length}</p>
        </div>
      </div>
      <div className="layout__content" data-testid={`${testId}-content`}>
        {children}
      </div>
    </>
  );
};
export default Layout;
