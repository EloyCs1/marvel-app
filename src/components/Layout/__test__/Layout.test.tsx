import { fireEvent, render } from "@testing-library/react";

import { AppWrapper } from "src/__test__/test.utils";
import { defaultFavoriteContextValue } from "src/context/FavoriteContext";
import * as FavoriteContext from "src/context/FavoriteContext";
import Layout from "../Layout";
import { CHARACTERS } from "src/mocks/mocks";
import { mockNavigate } from "src/setupTests";

describe("Layout", () => {
  test("Default render", () => {
    const mockValuesFavoriteProvider = {
      ...defaultFavoriteContextValue,
      favoriteIds: CHARACTERS.map(({ id }) => id),
      toggleFavoriteFilter: jest.fn(),
    };

    jest.spyOn(FavoriteContext, "useFavorites").mockReturnValue(mockValuesFavoriteProvider);

    const wrapper = render(
      <AppWrapper>
        <Layout>
          <div />
        </Layout>
      </AppWrapper>,
    );
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.getByTestId(`layout-count`).innerHTML).toBe(String(CHARACTERS.length));

    fireEvent.click(wrapper.getByTestId(`layout-icon-button`));
    fireEvent.click(wrapper.getByTestId(`layout-link`));
    expect(mockValuesFavoriteProvider.toggleFavoriteFilter).toBeCalledTimes(2);
    expect(mockNavigate).toBeCalled();
  });
});
