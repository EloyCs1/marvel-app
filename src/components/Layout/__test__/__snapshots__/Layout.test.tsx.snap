// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Default render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="navbar"
        data-testid="layout-navbar"
      >
        <a
          data-testid="layout-link"
          href="/"
        >
          <svg
            data-testid="layout-logo"
            fill="none"
            height="52"
            viewBox="0 0 130 52"
            width="130"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#clip0_14854_731)"
            >
              <path
                d="M130 0H0V52H130V0Z"
                fill="#EC1D24"
              />
              <path
                d="M126.222 40.059V47.965H111.58V4H119.465V40.059H126.222ZM63.658 25.559C63.048 25.853 62.41 25.999 61.788 26.001V11.861H61.828C62.45 11.856 67.092 12.045 67.092 18.854C67.092 22.413 65.512 24.658 63.658 25.559ZM40.55 34.24L42.733 15.441L44.998 34.24H40.55ZM110.205 12.025V4.007H87.879L84.204 30.786L80.574 4.006H72.522L73.423 11.156C72.495 9.324 69.199 4.006 61.943 4.006C61.896 4.004 53.883 4.006 53.883 4.006L53.852 43.038L47.984 4.007L37.439 4.002L31.367 44.442L31.369 4.007H21.278L17.64 26.724L14.096 4.006H4V47.972H11.95V26.78L15.568 47.972H19.794L23.359 26.78V47.972H38.686L39.614 41.21H45.784L46.711 47.972L61.758 47.98H61.768V47.972H61.788V33.702L63.633 33.432L67.45 47.982H75.234L75.232 47.972H75.254L70.243 30.924C72.781 29.044 75.649 24.28 74.886 19.721V19.719C74.894 19.777 79.615 48 79.615 48L88.871 47.973L95.198 8.123V47.973H110.205V40.065H103.081V29.985H110.205V21.955H103.081V12.024L110.205 12.025Z"
                fill="#FEFEFE"
              />
              <path
                d="M0 0H30V52H0V0Z"
                fill="#EC1D24"
              />
              <path
                d="M31.5 48V4H21.291L17.651 26.735L14.102 4H4V48H12V26.792L15.577 48H19.806L23.374 26.792V48H31.5Z"
                fill="#FEFEFE"
              />
            </g>
            <defs>
              <clippath
                id="clip0_14854_731"
              >
                <rect
                  fill="white"
                  height="52"
                  width="130"
                />
              </clippath>
            </defs>
          </svg>
        </a>
        <div
          class="navbar__favorites"
        >
          <button
            class="buttonIcon"
            data-testid="layout-icon-button"
          >
            <svg
              data-testid="layout-icon"
              fill="none"
              height="21.68"
              viewBox="0 0 24 22"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12 3.80348L6 0.161865L0 3.80348V11.607L12 21.8382L24 11.607V3.80348L18 0.161865L12 3.80348Z"
                fill="#EC1D24"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <p
            data-testid="layout-count"
          >
            4
          </p>
        </div>
      </div>
      <div
        class="layout__content"
        data-testid="layout-content"
      >
        <div />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="navbar"
      data-testid="layout-navbar"
    >
      <a
        data-testid="layout-link"
        href="/"
      >
        <svg
          data-testid="layout-logo"
          fill="none"
          height="52"
          viewBox="0 0 130 52"
          width="130"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            clip-path="url(#clip0_14854_731)"
          >
            <path
              d="M130 0H0V52H130V0Z"
              fill="#EC1D24"
            />
            <path
              d="M126.222 40.059V47.965H111.58V4H119.465V40.059H126.222ZM63.658 25.559C63.048 25.853 62.41 25.999 61.788 26.001V11.861H61.828C62.45 11.856 67.092 12.045 67.092 18.854C67.092 22.413 65.512 24.658 63.658 25.559ZM40.55 34.24L42.733 15.441L44.998 34.24H40.55ZM110.205 12.025V4.007H87.879L84.204 30.786L80.574 4.006H72.522L73.423 11.156C72.495 9.324 69.199 4.006 61.943 4.006C61.896 4.004 53.883 4.006 53.883 4.006L53.852 43.038L47.984 4.007L37.439 4.002L31.367 44.442L31.369 4.007H21.278L17.64 26.724L14.096 4.006H4V47.972H11.95V26.78L15.568 47.972H19.794L23.359 26.78V47.972H38.686L39.614 41.21H45.784L46.711 47.972L61.758 47.98H61.768V47.972H61.788V33.702L63.633 33.432L67.45 47.982H75.234L75.232 47.972H75.254L70.243 30.924C72.781 29.044 75.649 24.28 74.886 19.721V19.719C74.894 19.777 79.615 48 79.615 48L88.871 47.973L95.198 8.123V47.973H110.205V40.065H103.081V29.985H110.205V21.955H103.081V12.024L110.205 12.025Z"
              fill="#FEFEFE"
            />
            <path
              d="M0 0H30V52H0V0Z"
              fill="#EC1D24"
            />
            <path
              d="M31.5 48V4H21.291L17.651 26.735L14.102 4H4V48H12V26.792L15.577 48H19.806L23.374 26.792V48H31.5Z"
              fill="#FEFEFE"
            />
          </g>
          <defs>
            <clippath
              id="clip0_14854_731"
            >
              <rect
                fill="white"
                height="52"
                width="130"
              />
            </clippath>
          </defs>
        </svg>
      </a>
      <div
        class="navbar__favorites"
      >
        <button
          class="buttonIcon"
          data-testid="layout-icon-button"
        >
          <svg
            data-testid="layout-icon"
            fill="none"
            height="21.68"
            viewBox="0 0 24 22"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12 3.80348L6 0.161865L0 3.80348V11.607L12 21.8382L24 11.607V3.80348L18 0.161865L12 3.80348Z"
              fill="#EC1D24"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <p
          data-testid="layout-count"
        >
          4
        </p>
      </div>
    </div>
    <div
      class="layout__content"
      data-testid="layout-content"
    >
      <div />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
